### PR TITLE
feat(sumo): use perception via SUMOs context subscriptions

### PR DESCRIPTION
**Pull request by kschrab**
## Type of this PR 

- [ ] Bug fix
- [x] Enhancement

## Description

This PR adds the option to retrieve the surrounding vehicles by using context subscriptions with SUMO. 

Main feature: New `SumoPerceptionModule` sends a `VehicleSightDistanceConfiguration` if it is activated. SUMO creates context subscriptions for this vehicle with the defined range and opening angle. After sim-step, it reads context subscription results and adds a list of seen objects to the VehicleData. The `SumoPerceptionModule` transforms these objects to `VehicleObject` objects to be applicable with the API via `getPerceivedVehicles()`.

Further improvements/changes:
* Moved `PerformanceMonitor` to mosaic-utils to be usable in any part of MOSAIC
* Improved performance of `Grid`, mainy update method by reducing map calls.
* Removed internal follower-vehicle calculation; use subscriptions instead to retrieve follower.
* Added printing of all TraCI messages for debugging (optionally)

## Issue(s) related to this PR

 * Internal issue 303
 
## Definition of Done

- [x] You have read CONTRIBUTING.md carefully.
- [x] You have signed the [Contributor License Agreement](http://www.eclipse.org/legal/CLA.php).
- [x] All commits are signed-off (`-s`) with your mail address of your Eclipse Account.
- [x] `origin/main` has been merged into your Fork.
- [x] New functionality is covered by unit tests or integration tests. Code coverage must not decrease.
- [ ] There are no new SpotBugs warnings. 
- [x] Coding guidelines have been observed (see CONTRIBUTING.md).
- [x] All tests pass.

### CODE REVIEW

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/facades/SimulationFacade.java [hunk 1]
```diff
@@ -596,13 +643,18 @@ private VehicleSensors createSensorData(SumoVehicleState vehicleState, LeadingVe
...
+            VehicleData leadingVehicleData = getLastKnownVehicleData(leadingVehicle.getOtherVehicleId());
             leaderSpeed = leadingVehicleData != null ? leadingVehicleData.getSpeed() : -1;
         }
-        if (hasBackSensorActivated && followerDistance != null) {
-            rearDistance = followerDistance;
+        if (hasBackSensorActivated
+                && followerVehicle.getDistance() < vehicleState.rearSensorDistance
+        ) {
+            // using the own minGap is not 100% correct here, but it would be very expensive to fetch the minGap of the follower
+            rearDistance = followerVehicle.getDistance() + minGap;
```
Comment (schwepmo) [inline fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/facades/SimulationFacade.java:657]: not only is `minGap` inaccurate but wouldn't it make sense to measure from the back-bumper?

File: lib/mosaic-objects/src/main/java/org/eclipse/mosaic/lib/objects/vehicle/VehicleData.java [hunk 2]
```diff
@@ -303,6 +327,16 @@ public VehicleSensors getVehicleSensors() {
         return vehicleSensors;
     }
 
+    /**
+     * Getter for the list of vehicles in the field of view of this vehicle.
+     * Supported Traffic simulators: [SUMO (only with VehicleSightDistanceConfiguration, opt-in)]
+     *
+     * @return the list of vehicles in the field of view of this vehicle.
+     */
+    public List<SurroundingVehicle> getInSight() {
```
[the remark this answers is not in the document]
Comment (kschrab) [reply lib/mosaic-objects/src/main/java/org/eclipse/mosaic/lib/objects/vehicle/VehicleData.java:336]: sounds better, thanks!

File: fed/mosaic-application/src/test/java/org/eclipse/mosaic/fed/application/ambassador/UnitSimulatorTest.java [hunk 3]
```diff
@@ -76,6 +77,7 @@ public void setup() {
         registeredEvents.clear();
 
         SimulationKernel.SimulationKernel.setClassLoader(ClassLoader.getSystemClassLoader());
+        SimulationKernel.SimulationKernel.setConfiguration(new CApplicationAmbassador());
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-application/src/test/java/org/eclipse/mosaic/fed/application/ambassador/UnitSimulatorTest.java:80]: Not sure if there's a test which sets the config explicitly, which then would not work anymore as `setConfiguration` can only be called once. also I somehow like to see that the code under test requires a configuration object.

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/libsumo/SimulationSimulateStep.java [hunk 4]
```diff
@@ -141,22 +156,72 @@ private void readVehicles(List<AbstractSubscriptionResult> results) {
                 result.fuel = Vehicle.getFuelConsumption(sumoVehicleId);
             }
 
-            result.leadingVehicle = LeadingVehicle.NO_LEADER;
-
             if (fetchLeader) {
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/libsumo/SimulationSimulateStep.java]: done

**Review by rprotzmann — APPROVED** [review]

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/api/VehicleSubscribeSurroundingVehicle.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.eclipse.mosaic.fed.sumo.bridge.Bridge;
+import org.eclipse.mosaic.fed.sumo.bridge.CommandException;
+import org.eclipse.mosaic.rti.api.InternalFederateException;
+
+/**
+ * This class represents the SUMO command which allows to subscribe the vehicle to the application.
+ * Several options for vehicle subscription are implemented in this class.
+ */
+public interface VehicleSubscribeSurroundingVehicle {
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/api/VehicleSubscribeSurroundingVehicle.java:26]: All command classes in mosaic-sumo follow the scheme `EntityVerbObject` or  `EntityVerb`, e.g. `VehicleGetParameter`, `RouteGetEdges`, `VehicleAdd`, `VehicleSubscribe`. The current name matches this scheme. 

This also the reason why `VehicleSubscriptionSetFieldOfVision` is called as it is right now, as`VehicleSubscription` is the entity and `Set` the verb.

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/VehicleSubscriptionSetFieldOfVision.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+ * This class represents the SUMO command which allows to filter a previously created context subscription to
+ * collect all vehicles surrounding a specific vehicle.
+ */
+public class VehicleSubscriptionSetFieldOfVision
+        extends AbstractTraciCommand<Void>
+        implements org.eclipse.mosaic.fed.sumo.bridge.api.VehicleSubscriptionSetFieldOfVision {
+
+    @SuppressWarnings("WeakerAccess")
+    public VehicleSubscriptionSetFieldOfVision() {
+        super(SumoVersion.SUMO_1_4_x);
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/bridge/traci/VehicleSubscriptionSetFieldOfVision.java:35]: We use `SumoVersion.SUMO_1_4_x` one here to define, that this TraCI command only works with SUMO versions of at least 1.4.0. If an older version of SUMO is installed, this command is not executed and a warning is logged instead.

File: fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java [hunk 7]
```diff
@@ -161,18 +161,17 @@ private void receiveInteraction(VehicleRegistration interaction) {
...
+            logMessage = "VehicleRegistration from RTI \"{}\" received at simulation time {} ns (subscribe={})";
         } else { // still subscribe to vehicles with apps
-            logMessage = "VehicleRegistration for SUMO vehicle \"{}\" received at simulation time {} ns";
+            logMessage = "VehicleRegistration for SUMO vehicle \"{}\" received at simulation time {} ns (subscribe={})";
         }
-        log.info(logMessage, vehicleId, interaction.getTime());
 
-        // now prepare vehicles to subscribe to
         boolean subscribeToVehicle = sumoConfig.subscribeToAllVehicles || vehicleMapping.hasApplication();
+        log.info(logMessage, vehicleId, interaction.getTime(), subscribeToVehicle);
```
[the remark this answers is not in the document]
Comment (kschrab) [reply fed/mosaic-sumo/src/main/java/org/eclipse/mosaic/fed/sumo/ambassador/SumoAmbassador.java:170]: I'll fix that

**Pull request merged**